stablehlo/reference/Ops.cpp

```diff
@@ -897,6 +933,49 @@ Tensor evalFloorOp(const Tensor &operand, ShapedType resultType) {
   return result;
 }
 
+Tensor evalGatherOp(const Tensor &operand, const Tensor &startIndices,
+                    const Axes &offsetDims, const Axes &collapsedSliceDims,
+                    const Axes &startIndexMap, Axis indexVectorDim,
+                    const Sizes &sliceSizes, bool indicesAreSorted,
+                    ShapedType resultType) {
+  Tensor result(resultType);
+  Axes batchDims;
+  for (auto d : result.getAxes())
+    if (!llvm::is_contained(offsetDims, d)) batchDims.push_back(d);
+
+  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+       ++resultIt) {
+    Index batchIndex;
+    for (auto d : batchDims) batchIndex.push_back((*resultIt)[d]);
+
+    Index startIndicesIndex = batchIndex;
+    if (indexVectorDim < startIndices.getRank())
+      startIndicesIndex.insert(startIndicesIndex.begin() + indexVectorDim,
+                               kColon);
+    Index startIndex = evalIndex(evalSliceOp(startIndices, startIndicesIndex));
+
+    Index fullStartIndex(operand.getRank(), 0);
+    for (auto dOperand : operand.getAxes()) {
+      if (llvm::find(startIndexMap, dOperand) == startIndexMap.end()) continue;
+      auto dStart = llvm::find(startIndexMap, dOperand) - startIndexMap.begin();
+      fullStartIndex[dOperand] = startIndex[dStart];
+    }
+
+    Index offsetIndex;
+    for (auto d : offsetDims) offsetIndex.push_back((*resultIt)[d]);
+
+    Index fullOffsetIndex(offsetIndex.size() + collapsedSliceDims.size(), 0);
+    for (size_t i = 0, oi = 0; i < fullOffsetIndex.size(); ++i) {
+      if (llvm::is_contained(collapsedSliceDims, i)) continue;
+      fullOffsetIndex[i] = offsetIndex[oi++];
+    }
+
+    Index operandIndex = fullStartIndex + fullOffsetIndex;
+    result.set(*resultIt, operand.get(operandIndex));
```

Comment:
Let's guard with `if (operandIndex.inBounds(operand.getShape())` to avoid outright crashes?